gemrb/core/Scriptable/Actor.cpp

```diff
@@ -8184,14 +8184,36 @@ void Actor::AddExperience(int exp, int combat)
 		bonus += adjustmentPercent;
 	}
 	bonus += GetFavoredPenalties();
-	exp = ((exp * (100 + bonus)) / 100) + BaseStats[IE_XP];
+
+	int classid = GetActiveClass() -1; //-1 to allow lookup in xpcap array
+	int xp_which = IE_XP;
+
+	// decide which particular XP stat to add to (only for TNO's switchable classes)
+	if (pstflags) {
+		Game* game = core->GetGame();
+		// rule only applies to the protagonist
+		if ( GetScriptName() == game->GetPC(0, false)->GetScriptName() ) {
```

Comment:
You can compare the pointers directly, no need to look up the script name.

Reply:
and you can do it one level higher, so there's less nesting.

Reply:
I was basically looking around for anywhere else that checks if the actor is TNO, and that's the first example I found:

https://github.com/gemrb/gemrb/blob/d5da8d109a65d6b55e1ae4f660c9e6a7d213492f/gemrb/core/Scriptable/Actor.cpp#L5849-L5853

Not sure if there is a less convoluted check somewhere else, if not, should I come up with one?

Reply:
It's not convoluted, it can just be improved (which will also make it work in iwds). You can add a `Game::IsProtagonist(Actor *pc)` method to abstract it away.

Reply:
Out of curiosity, is there any reason not to just use "this" to compare to PC 0 ? I pushed a modified example that seems to work fine

Reply:
Also, I think IWD is fine in the lines I linked to, because PM_NO is only set for PST, IWD gets PM_TEAM

Reply:
this: that's what I meant by the first comment.
iwd: the function wouldn't check for protagonist mode though, so it would break if it checked for script name.

Reply:
I'm not sure I'm following 100%

I'm not really trying to add a check to detect the protagonist, I have to specifically identify the Nameless one is the actor here, since he is the only character with this XP gain style. That's what I meant the `pstflags` part of the check to achieve.

According to IESDP, the other games don't even have these extra stats, not even IWD2. They keep track of the different levels, sure, but the XP is one stat divvied between number of classes (or whatever other method for dual class chars)

Unless we're talking about completely different things?

Reply:
I was talking about a Game::IsProtagonist.

Reply:
Ah right, I'm not even sure how I would decide that in IWD, I thought it wasn't even supposed to treat any one character as the 'main' character. 

But again, the intention is just to isolate this block to PST and TNO specifically, it shouldn't go near any of the other games actors, or even the other PST npc's which already work fine.

Reply:
The protagonist is always the first character, that's what the check is about.